src/data/apigateway/mod.rs

```diff
@@ -0,0 +1,170 @@
+    HEAD,
+    GET,
+    POST,
+    PUT,
+    OPTIONS,
+    DELETE,
+}
+
+impl fmt::Display for HttpMethod {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", match *self {
+            HttpMethod::HEAD => "HEAD",
+            HttpMethod::GET => "GET",
+            HttpMethod::POST => "POST",
+            HttpMethod::PUT => "PUT",
+            HttpMethod::OPTIONS => "OPTIONS",
+            HttpMethod::DELETE => "DELETE"
+        })
+    }
+}
+
+#[derive(Serialize,Deserialize)]
+pub struct HttpEvent {
+    pub body: Option<String>,
+    pub headers: Option<HashMap<String, String>>,
+    #[serde(rename="httpMethod")]
+    pub http_method: HttpMethod,
+    #[serde(rename="isBase64Encoded")]
+    pub is_base64_encoded: bool,
+    pub path: String,
+    #[serde(rename="pathParameters")]
+    pub path_parameters: Option<HashMap<String, String>>,
+    #[serde(rename="queryStringParameters")]
+    pub query_string_parameters: Option<HashMap<String, String>>,
+    pub resource: String,
+    #[serde(rename="requestContext")]
+    pub request_context: HttpEventRequestContext,
+    #[serde(rename="stageVariables")]
+    pub stage_variables: Option<HashMap<String, String>>,
+}
+
+impl HttpEvent {
+
+    pub fn get_header(&self, key: &str) -> Option<&str> {
+        match self.headers {
+            Some(ref h) => h.get(key).map(|s| s.as_str()),
+            None => None,
+        }
+    }
+}
+
+impl fmt::Display for HttpEvent {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{protocol} {method} {path}{querystring} (body? {body})",
+            protocol = match self.request_context.protocol {
+                Some(ref s) => s.as_str(),
+                None => "(unknown)"
+            },
+            method = self.http_method,
+            path = self.path,
+            querystring = qs::to_string(&self.query_string_parameters).unwrap_or(String::new()),
+            body = match self.body {
+                Some(_) => true,
+                None => false
+            }
+        )
+    }
+}
+
+#[derive(Serialize,Deserialize)]
+pub struct HttpEventRequestContext {
+    #[serde(rename="accountId")]
+    pub account_id: String,
+    #[serde(rename="apiId")]
+    pub api_id: String,
+    #[serde(rename="httpMethod")]
+    pub http_method: HttpMethod,
+    pub identity: HashMap<String, Option<String>>,
+    pub path: String,
+    pub protocol: Option<String>,
+    #[serde(rename="requestId")]
+    pub request_id: String,
+    #[serde(rename="requestTime")]
+    pub request_time: Option<String>,
+    #[serde(rename="requestTimeEpoch")]
+    pub request_time_epoch: Option<u64>,
+    #[serde(rename="resourceId")]
+    pub resource_id: String,
+    #[serde(rename="resourcePath")]
+    pub resource_path: String,
+    pub stage: String,
+}
+
+impl HttpEventRequestContext {
+
+    pub fn time(&self) -> Option<DateTime<Utc>> {
+        // Utc::datetime_from_str(&self.request_time, "").ok()e
```

Comment:
you could probably ditch a hard coupling of crowbar to a [crate](https://github.com/ilianaw/rust-crowbar/pull/31/files#diff-80398c5faae3c069e4e6aa2ed11b28c0R17) by letting it be a user decision to select a time parsing crate and let crowbar be smaller.

Reply:
How can we genericize this to work like that?

Reply:
If we didn't want to represent this as a string that lambda events do, you could maybe use a feature + cfg. I don't feel really strongly about this. Only that I'd like to see crowbar have a solve only the minimal set of problems with a minimal set of dependencies. Motivations are that it lets users pick a crate they prefer for representing time typically so that they only have one crate for dealing with time in their projects ( think about what hyper did with tls ) and it frees up crowbar up from a bit of maintenance by not worrying falling behind on its dependencies. I also don't feel too strongly here but that's my two cents. I'd feel otherwise if rust had a std lib way of representing timestamps.

Reply:
I don't think genericizing the time crate is necessary to land this PR.